lib/charms/s3_integrator/v0/s3.py

```diff
@@ -0,0 +1,479 @@
+        pass
+
+ if __name__ == "__main__":
+    main(ExampleRequirerCharm)
+```
+
+"""
+import logging
+from typing import Dict, List, Optional
+
+import ops.charm
+import ops.framework
+import ops.model
+from ops.charm import (
+    EventSource,
+    Object,
+    ObjectEvents,
+    RelationBrokenEvent,
+    RelationChangedEvent,
+    RelationEvent,
+)
+from ops.model import Relation
+
+LIBID = "55c4cfd4a8ac45b1b62f2826272a1cf8"
+LIBAPI = 0
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class CredentialRequestedEvent(RelationEvent):
+    """Event emitted when a set of credential is requested for use on this relation."""
+
+
+class S3CredentialEvents(ObjectEvents):
+    """Event descriptor for events raised by S3Provider."""
+
+    credentials_requested = EventSource(CredentialRequestedEvent)
+
+
+class S3Provider(Object):
+    """A provider handler for communicating S3 credentials to consumers."""
+
+    on = S3CredentialEvents()
+
+    def __init__(
+        self,
+        charm: ops.charm.CharmBase,
+        relation_name: str,
+    ):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.local_app = self.charm.model.app
+        self.local_unit = self.charm.unit
+        self.relation_name = relation_name
+
+        # monitor relation changed event for changes in the credentials
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """React to the relation joined event by consuming data."""
+        if not self.charm.unit.is_leader():
+            return
+        # emit on credential request event
+        self.on.credentials_requested.emit(event.relation, app=event.app, unit=event.unit)
+
+    def fetch_relation_data(self) -> dict:
+        """Retrieves data from relation.
+
+        This function can be used to retrieve data from a relation
+        in the charm code when outside an event callback.
+
+        Returns:
+            a dict of the values stored in the relation data bag
+                for all relation instances (indexed by the relation id).
+        """
+        data = {}
+
+        for relation in self.relations:
+            data[relation.id] = {
+                key: value for key, value in relation.data[self.charm.app].items()
+            }
+        return data
+
+    def update_credential_data(self, relation_id: int, data: dict) -> None:
+        """Updates the credential data as set of key-value pairs in the relation.
+
+        This function writes in the application data bag, therefore,
+        only the leader unit can call it.
+
+        Args:
+            relation_id: the identifier for a particular relation.
+            data: dict containing the key-value pairs
+                that should be updated.
+        """
+        # check and write changes only if you are the leader
+        if self.local_unit.is_leader():
```

Comment:
reduce indentation complexity 

```suggestion
        # check and write changes only if you are the leader
        if not self.local_unit.is_leader():
             return
```